glengine/gl_context.cpp

```diff
@@ -1,22 +1,178 @@
 #include "gl_context.h"
 
+#include "glad/glad.h"
+#define GLFW_INCLUDE_NONE
+#include <GLFW/glfw3.h>
+
+#include "math/vmath.h"
+#include "math/math_utils.h"
+
 #include "imgui/imgui.h"
 #include "imgui/imgui_impl_glfw.h"
-// #include "imgui/imgui_impl_opengl3.h"
+
+#include "stb/stb_image_write.h"
+#include "gl_engine.h"
 
 #include <cstdlib>
 #include <cstdio>
 
+namespace glengine {
+
+// input states that are needed across different callbacks
+struct InputState {
+    bool left_button_pressed = false;
+    bool middle_button_pressed = false;
+    bool right_button_pressed = false;
+    bool ctrl_key_pressed = false;
+    bool shift_key_pressed = false;
+    bool alt_key_pressed = false;
+    math::Vector2i previous_cursor_pos{-1,-1};
+};
+
+// information about the current window
+struct WindowState {
+    math::Vector2i window_size{-1,-1};
+    math::Vector2i framebuffer_size{-1,-1};
+};
+
+struct Context {
+    GLFWwindow *window = nullptr;
+    InputState input_state;
+    WindowState window_state;
+};
+
+} //
+
+namespace {
+
+// global application context
+glengine::Context g_context;
```

Comment:
not the most elegant solution but gets the job done; I prefer this for now instead of having something like a `void*` pointer in the glengine class.